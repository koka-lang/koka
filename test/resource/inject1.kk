// Test inject

fun foo(action : () -> e a ) : <ndet|e> a
{
  val x = inject<ndet>(action)
  if (random-bool()) then foo(action) else x 
}

fun main() {
  foo{42}.println
}