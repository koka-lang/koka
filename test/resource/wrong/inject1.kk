// Test inject over handled effects

effect heap<a> {
  fun get() : a 
  fun set(value : a) : ()
}

val new-ref = handler(s) {
  get() -> resume(s,s)
  set(x) -> resume((),x)
}

fun main() {
  new-ref("hi") {
    new-ref(44) {    
      get() - inject<int>{(get():string).count}  // 44 - "hi".count
    }
  }
}