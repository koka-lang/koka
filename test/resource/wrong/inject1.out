test/algeff/wrong/inject1.kk(16,15): error: Invalid type
 type context : inject<int>{(get():string).count}
 type : inject<int>{(get():string).count}
 inferred kind: V
 expected kind: X
 because : Can only inject effect constants (of kind X)